[clang] Reject 'auto' storage class with type specifier in C++

osamakader opened the pull request:
Fixes #164273

shafik commented:
I am going to be at the wg21 meeting next week, so I may not have a chance to look at this.

Endilll commented on clang/test/CXX/drs/cwg3xx.cpp:
```diff
@@ -1732,11 +1732,13 @@ namespace cwg395 { // cwg395: 3.0
-    // since-cxx11-error@-1 {{'auto' storage class specifier is not permitted in C++11, and will not be supported in future releases}}
-    // expected-error@-2 {{illegal storage class on function}}
+    // since-cxx11-error@-1 {{'auto' cannot be combined with a type specifier in C++}}
+    // cxx98-error@-2 {{illegal storage class on function}}
```
I'd prefer to see C++98 diagnostic first. This way expected directives are in chronological order.

osamakader replied:
OK

Endilll reviewed (commented):
Changes to C++ DR tests LGTM otherwise.

zygoloid commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4095,11 +4095,22 @@ void Parser::ParseDeclarationSpecifiers(
-              << FixItHint::CreateRemoval(DS.getStorageClassSpecLoc());
+          // In C++ (not C23), 'auto' cannot be combined with a type specifier.
+          // However, OpenCL has its own error handling for this case.
+          if (getLangOpts().CPlusPlus && !getLangOpts().C23 &&
```
No need to check both CPlusPlus and C23 here; they're mutually exclusive.

zygoloid commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4095,11 +4095,22 @@ void Parser::ParseDeclarationSpecifiers(
-            Diag(Tok, diag::ext_auto_storage_class)
-              << FixItHint::CreateRemoval(DS.getStorageClassSpecLoc());
+          // In C++ (not C23), 'auto' cannot be combined with a type specifier.
+          // However, OpenCL has its own error handling for this case.
```
What is the OpenCL handling? If it's the same as the general C++ handling, can we remove the OpenCL handling and enable the new code for OpenCL too?

svenvh replied:
Yes, I think we can drop the OpenCL specific diagnostic in favor of the C++ diagnostic.

Sirraide commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -399,6 +399,8 @@ def err_requires_clause_on_declarator_not_declaring_a_function : Error<
+def err_auto_type_specifier : Error<
+  "'auto' cannot be combined with a type specifier">;
 def ext_auto_storage_class : ExtWarn<
   "'auto' storage class specifier is not permitted in C++11, and will not "
```
This pr makes `ext_auto_storage_class` unused, so we should remove it

AaronBallman commented on clang/test/Parser/cxx-auto-type-specifier.cpp:
```diff
@@ -0,0 +1,22 @@
+
+// Test that 'auto' cannot be combined with a type specifier in C++.
+void f() {
+  auto int x = 1;        // expected-error {{'auto' cannot be combined with a type specifier}}
```
One more test to add:
```
template <typename Ty>
void g() {
  auto Ty x;
}

void test() {
  g<float>();
}
```

osamakader replied:
@AaronBallman

AaronBallman replied:
This might not be an easy thing to fix. One thing to try investigating is: we have this `annot_typename` token that is an annotation token saying "the token is a type". This is produced from calling one of the `TryAnnotate*()` functions in the parser; that basically does some tentative parsing and other tricks to potentially produce the annotation token so when the parser gets the next token, it gets the annotation that did all of the heavy lifting. What I'd try is to see if we can get a `TryAnnotate*()` function to replace the call to `GetLookAheadToken(1)`. The check for the type specifier will work correctly with an `annot_typename` token, but none of the existing try-annotation functions work off a given token, they all start from the current token.

Does that make sense?

osamakader replied:
So, I tried that:

```
     TentativeParsingAction TPA(*this);
     ConsumeToken(); 
     TryAnnotateTypeOrScopeToken()
     NextIsTypeSpecifier = isKnownToBeTypeSpecifier(Tok);
     TPA.Revert();
```

TryAnnotate* incorrectly annotates some syntax e.g. in llvm-project/clang/test/Parser/cxx-concepts-requires-clause.cpp:

```
template <typename T> T Foo;
template <typename T> auto C(Foo<T>);
```

As true which breaks this test and many other tests.

mizvekov replied:
Yeah, I don't see what annotation we would need. For this example:
```
template <typename Ty>
void g() {
  auto Ty x;
}
```

We know `Ty` is a type, and we know there is no type which could be substituted here making this definition valid, so we can diagnose this right away.

osamakader replied:
Latest commit has that template test case passing without annotation.

zygoloid replied:
You don't need to diagnose conflicts from `Finish`. Here's one approach:

* When consuming `auto`, treat it as specified by the underlying language (either a type specifier or a storage class specifier).
* If we encounter a conflict between storage classes, if one of them is `auto` and there is no type specifier yet, then set the type specifier to `auto` and keep the other storage class as the "real" storage class specifier.
* In `Finish`, if there is a storage class of `auto` and no type specifier at all, and we're in a language mode without "implicit int", then convert the storage class to a type specifier.

Finally, if we have a type specifier of `TST_auto` in a language that doesn't support it, then produce a warning.

In fact, looking at the `DeclSpec` implementation, we already do all of that. So I'm a bit confused about what we're discussing.

AaronBallman commented on clang/test/CXX/dcl.dcl/dcl.spec/dcl.type/dcl.spec.auto/p3.cpp:
```diff
@@ -42,7 +42,12 @@ void p3example() {
   static auto y = 0.0;
```
> I see a diagnostic for it appearing in all of your test cases:

Ugh, yeah, I am seeing that now, I think I was looking at the wrong output before. :-D

Yeah, I think the table just needs to be updated, that doesn't need to be a part of this PR though. Thanks!

osamakader committed:
Fix handling of 'auto' type specifier conflicts

This commit fixes several issues related to 'auto' type specifier handling.

Core implementation:
- Add ConflictingTypeSpecifier flag to DeclSpec to track when 'auto' conflicts
  with another type specifier (e.g., 'auto int' or 'int auto'). This allows
  Finish() to properly handle the conflict based on language mode rather than
  immediately emitting an error during parsing.

1. Fix 'auto auto' duplicate detection: Keep TypeSpecType as TST_auto after
   emitting error so subsequent checks can emit function-specific errors
   (e.g., 'not allowed in function prototype').

2. Fix OpenCL 'auto' and 'register' storage class errors: Reorder checks to
   prioritize OpenCL-specific handling before C++11+ checks. Use
   SetStorageClassSpec to trigger OpenCL validation and emit version-specific
   error messages.

3. Fix C23 'int auto' and 'constexpr int auto' handling:
   - For 'constexpr int auto': Treat 'auto' as type specifier conflict (don't
     convert to storage class), emit single error message.
   - For 'int auto' without constexpr: Convert 'auto' to storage class specifier
     (no type conflict error), only emit 'illegal storage class' if applicable.

4. Fix template parameter parsing: Skip type name lookup when 'auto' is set
   in template parameter context to avoid false ambiguity errors for
   placeholder variables.

5. Fix concept constraint syntax: Detect 'C<T> auto' pattern and skip type
   conflict detection for concept constraints.

6. Update test expectations: Align error messages and line numbers with
   actual diagnostic locations and behavior.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Fix handling of 'auto' type specifier conflicts

This commit fixes several issues related to 'auto' type specifier handling.

Core implementation:
- Add ConflictingTypeSpecifier flag to DeclSpec to track when 'auto' conflicts
  with another type specifier (e.g., 'auto int' or 'int auto'). This allows
  Finish() to properly handle the conflict based on language mode rather than
  immediately emitting an error during parsing.

1. Fix 'auto auto' duplicate detection: Keep TypeSpecType as TST_auto after
   emitting error so subsequent checks can emit function-specific errors
   (e.g., 'not allowed in function prototype').

2. Fix OpenCL 'auto' and 'register' storage class errors: Reorder checks to
   prioritize OpenCL-specific handling before C++11+ checks. Use
   SetStorageClassSpec to trigger OpenCL validation and emit version-specific
   error messages.

3. Fix C23 'int auto' and 'constexpr int auto' handling:
   - For 'constexpr int auto': Treat 'auto' as type specifier conflict (don't
     convert to storage class), emit single error message.
   - For 'int auto' without constexpr: Convert 'auto' to storage class specifier
     (no type conflict error), only emit 'illegal storage class' if applicable.

4. Fix template parameter parsing: Skip type name lookup when 'auto' is set
   in template parameter context to avoid false ambiguity errors for
   placeholder variables.

5. Fix concept constraint syntax: Detect 'C<T> auto' pattern and skip type
   conflict detection for concept constraints.

6. Update test expectations: Align error messages and line numbers with
   actual diagnostic locations and behavior.

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader commented:
ping :)

AaronBallman commented on clang/include/clang/Sema/DeclSpec.h:
```diff
@@ -415,6 +415,12 @@ class DeclSpec {
+  // Track conflicting type specifier when 'auto' is set (for Finish()
+  // detection)
+  LLVM_PREFERRED_TYPE(TST)
+  unsigned ConflictingTypeSpecifier : 7;
```
This should be moved up so it's grouped with the other bit-fields, otherwise it won't pack together. I'd put it up with the other type-specifier fields.

AaronBallman commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -3861,6 +3876,47 @@ void Parser::ParseDeclarationSpecifiers(
+            NextKind == tok::coloncolon || NextKind == tok::comma ||
+            NextKind == tok::semi || NextKind == tok::colon ||
+            NextKind == tok::greater || NextKind == tok::r_paren ||
+            NextKind == tok::arrow) {
```
```suggestion
        Token Next = NextToken();
        if (Next.isOneOf(tok::equal, tok::l_paren, tok::l_square, ...)) {
```
I think this will be a cleaner way to express the list.

AaronBallman reviewed (commented):
(Some feedback, I've not had the chance to do a thorough review of all the parsing changes though.)

osamakader committed:
Add more auto type tests, and some refactoring

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Add more auto type tests, and some refactoring

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Fix DeclSpec merge conflict

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Fix DeclSpec merge conflict

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Fix format error

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Fix format error

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Reordering constructor initializers to match member declaration order

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Reordering constructor initializers to match member declaration order

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader commented on clang/test/SemaCXX/auto-cxx0x.cpp:
```diff
@@ -1,8 +1,8 @@
-  auto int a; // expected-warning {{'auto' storage class specifier is not permitted in C++11, and will not be supported in future releases}}
-  int auto b; // expected-error{{cannot combine with previous 'int' declaration specifier}}
+  auto int a; // expected-error {{'auto' cannot be combined with a type specifier}}
+  int auto b; // expected-error {{'auto' cannot be combined with a type specifier}}
```
sorry for the late reply, done.

to268 commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -3860,6 +3875,44 @@ void Parser::ParseDeclarationSpecifiers(
+          }
+        }
+      }
+
```
I would rather move this part in Sema if possible.

osamakader replied:
I moved the auto conflict handling/diagnostic into DeclSpec::Finish() in Sema. I think this remaining part needs to stay in the parser because it decides where declaration-specifiers end before consuming the identifier. Once the parser consumes a typedef-name-looking identifier as a type specifier, Sema can diagnose the conflict but cannot recover the intended declarator-id parse for cases like [](auto c), template<auto V>, or for (auto Arg : range).

to268 commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -4093,15 +4146,33 @@ void Parser::ParseDeclarationSpecifiers(
+        // Always set 'auto' as a type specifier in C++11+ and C23.
+        // Conflicts will be detected in DeclSpec::Finish().
+        isInvalid = DS.SetTypeSpecType(DeclSpec::TST_auto, Loc, PrevSpec,
+                                       DiagID, Policy);
```
> I would rather move this part in Sema if possible.

Same for here.

osamakader replied:
removed from parser.

to268 reviewed (commented):
I would like to move some checks as much as possible from the Parsing stage to the Sema stage primarily for maintainability reasons, as we probably have code that parses templates and concept constraints already, and it would be easier to acquire the "parts" you need.

I also would like to see this patch rebased on `main` due to the following commits:
- https://github.com/llvm/llvm-project/commit/211b88b64b0e409a2e294a1fe444b305aa94a2a6
- https://github.com/llvm/llvm-project/commit/42b6a6faaa0014f7e39c57c6221fec1855ec4409

shafik commented:
ping

Endilll commented on clang/test/CXX/drs/cwg3xx.cpp:
```diff
@@ -1704,13 +1704,23 @@ namespace cwg395 { // cwg395: 3.0
-    // since-cxx11-error@-1 {{'auto' storage class specifier is not permitted in C++11, and will not be supported in future releases}}
-    // expected-error@-2 {{illegal storage class on function}}
+#if __cplusplus >= 201103L
+    // expected-error@-2 {{'auto' cannot be combined with a type specifier}}
```
```suggestion
    // since-cxx11-error@-1 {{'auto' cannot be combined with a type specifier}}
```
No need to use `#if __cplusplus` for expected directives.

osamakader replied:
Done.

osamakader commented:
I'm going to address the recent reviews ASAP, sorry was a bit sick.

osamakader committed:
clean up ParseDecl auto type part

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
clean up ParseDecl auto type part

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
use since-cxx11-error for auto type in cwg3xx.cpp test

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
use since-cxx11-error for auto type in cwg3xx.cpp test

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader commented:
The PR is ready for review.

to268 commented on clang/lib/Sema/DeclSpec.cpp:
```diff
@@ -1198,6 +1242,164 @@ void DeclSpec::Finish(Sema &S, const PrintingPolicy &Policy) {
+        ConflictingTypeSpecifierLoc = SourceLocation();
+      } else if (ConflictingTypeSpecifier == TST_auto) {
+        // "int auto" case: Convert 'auto' to storage class specifier
+        // In C23, if 'constexpr' is present, treat 'auto' as a type specifier
```
TODO: Is it only a type specifier when the `constexpr` keyword is involved in C23?

shafik commented on clang/lib/Sema/DeclSpec.cpp:
```diff
@@ -759,6 +759,17 @@ bool DeclSpec::SetTypeSpecType(TST T, SourceLocation TagKwLoc,
   if (TypeSpecType == TST_error)
     return false;
   if (TypeSpecType != TST_unspecified) {
+    // Store conflicting type specifier for Finish() to detect:
```
This code is repeated four times, this is a maintainability problem. Identical code segments tend to drift over time generating more bugs and uneven updates. 

Please factor out in some way.

osamakader replied:
Done.

shafik commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -3775,7 +3775,9 @@ void Parser::ParseDeclarationSpecifiers(
-      if (DS.hasTypeSpecifier())
+      // However, if 'auto' is set, we need to check if this identifier is a
+      // type name to detect conflicts (e.g., "auto MyInt").
+      if (DS.hasTypeSpecifier() && DS.getTypeSpecType() != DeclSpec::TST_auto)
```
It is not clear to me what function these changes in `ParseDeclarationSpecifiers` have. Can you explain? 

This should really be explained in the summary, what are you changing and why.

osamakader replied:
This parser change exists to let Sema see auto combined with a following typedef-name type specifier.

Previously, once auto was already parsed as a type specifier, this early exit stopped parsing declaration specifiers before looking up the next identifier. That meant cases like
```
typedef int MyInt;
auto MyInt x;
```
would stop at MyInt as if it were the declarator-id, instead of recognizing MyInt as a second type specifier and letting DeclSpec::Finish() diagnose auto plus another type specifier.

shafik reviewed (commented):
Please also add a release note.

to268 reviewed (commented):
I would like to see some tests cases on the C side (for both pre-C23 and C23).

osamakader committed:
refactor conflicting type tracking

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
refactor conflicting type tracking

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
add c2x auto with typedef specifier

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
add c2x auto with typedef specifier

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
add release note - reject auto with type spec in c++

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
add release note - reject auto with type spec in c++

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>

osamakader committed:
Merge branch 'main' into fix-auto-with-type

osamakader commented:
> I would like to see some tests cases on the C side (for both pre-C23 and C23).

Done.

to268 reviewed (approved):
LGTM on the way it is implemented + tested with C23's `auto`.

osamakader committed:
update clang/test/CXX/dcl/dcl.fct/p17.cpp expectation

Signed-off-by: Osama Abdelkader <osama.abdelkader@gmail.com>
